### PR TITLE
[ECO-5178] Fix sending and receiving of message headers and metadata

### LINKED ISSUE #193: Sending a message with metadata or headers doesn't work
In the integration tests, try adding the following to a call to `room.messages.send`:

```swift
metadata: ["hello": .number(12), "goodbye": .string("what")]
```

You'll see that the call to `send` fails with `noItemInResponse`, and in the logs we see

```
Could not messagepack object: AblyChat.MetadataValue.number(12)
Could not messagepack object: AblyChat.MetadataValue.string("what")
WARN: (ARTHttp.m:57) unable to publish chat message; cannot decode request body
```

Similarly, with headers:

```swift
headers: ["hello": .number(12), "goodbye": .string("what")]
```

we get

```
Could not messagepack object: AblyChat.HeadersValue.string("what")
Could not messagepack object: AblyChat.HeadersValue.number(12)
WARN: (ARTHttp.m:57) unable to publish chat message; cannot decode request body
```

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5178) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
Resolves #193. See commit messages for more details.

## Summary by CodeRabbit

- **New Features**
	- Enhanced JSON handling for message and presence data, improving type safety and error handling.
	- Added structured data extraction for incoming messages, metadata, and headers.

- **Bug Fixes**
	- Improved error messages for missing or malformed data in message handling.

- **Tests**
	- Added new tests for validating message sending and subscription functionalities.
	- Updated existing tests to reflect changes in method names and improve clarity.

- **Documentation**
	- Updated comments and method signatures for consistency and clarity in JSON handling.

### CODE REVIEW

File: Sources/AblyChat/JSONCodable.swift [hunk 1]
```diff
@@ -0,0 +1,9 @@
+internal protocol JSONEncodable {
+    var toJSONObjectValue: [String: JSONValue] { get }
+}
+
+internal protocol JSONDecodable {
+    init(jsonValue: JSONValue) throws
+}
+
+internal typealias JSONCodable = JSONDecodable & JSONEncodable
```
Comment (maratal) [inline Sources/AblyChat/JSONCodable.swift:9]: Why couldn't we just use `Codable` as presence user data?

Comment (lawrence-forooghian) [reply Sources/AblyChat/JSONCodable.swift:9]: The way you'd usually use `Codable` is with `JSONEncoder` which converts to and from `Data`. We want to convert to and from objects that can be passed to ably-cocoa such as `Dictionary` etc.

File: Tests/AblyChatTests/ChatAPITests.swift [hunk 2]
```diff
@@ -52,6 +52,52 @@ struct ChatAPITests {
...
+        }
+        let chatAPI = ChatAPI(realtime: realtime)
+
+        // When
+        _ = try await chatAPI.sendMessage(
+            roomId: "", // arbitrary
+            params: .init(
+                text: "", // arbitrary
+                // The exact value here is arbitrary, just want to check it gets serialized
+                headers: ["numberKey": .number(10), "stringKey": .string("hello")]
```
Comment (maratal) [inline Tests/AblyChatTests/ChatAPITests.swift:69]: Why can't it be ["numberKey": 10, "stringKey": "hello"], where dictionary is [`String`, `AnyPrimitiveType`], where `AnyPrimitiveType` is just an empty protocol, to which `String`, `Int` etc are conformed?

Comment (maratal) [reply Tests/AblyChatTests/ChatAPITests.swift:69]: ok, this will not work for nested dictionaries

**Review by maratal — APPROVED** [review]
LGTM

**Pull request merged**